### ISSUE TITLE
[question] AutotoolsToolchain uses MSVC flags with clang

### ISSUE THREAD

**Issue by jijjijj** [conan-io/conan#17342]
### What is your question?

So I have a Win64 build with `clang` set as a compiler and I try to build `mpg123` which uses Autotools to configure itself. The build fails because Autotools receive `-MDd` flag inside `CFLAGS` which is not supported by `clang/clang++` only by `clang-cl` (which I don't want to use).

So I went to code of AutotoolsToolchain and I found this
![image](https://github.com/user-attachments/assets/b68c3ec8-ead5-4e27-b3a4-a5ceb11d39a4)
Where `_get_msvc_runtime_flag()` returns the runtime flag appropriate for `MS cl.exe` or `clang-cl` but not for clang. And then in `cflags()` it just returns it without any regard to the actual compiler in use.

At the same time, when choosing flags relevant for mac/apple compilers it does filter the flags out in case the current system isn't appropriate.
![image](https://github.com/user-attachments/assets/afede3dd-0395-4409-955b-3e75dc0a22db)

What's up with that?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @jijjijj 

Thanks for your question.

>  Autotools receive -MDd flag inside CFLAGS which is not supported by clang/clang++ only by clang-cl (which I don't want to use).

Could  you please clarify which clang are you using in Windows, the msvc-backed clang (irrespective of clang or clang-cl front), or the msys2-backed clang?

In case you haven't read this, this blog post is relevant: https://blog.conan.io/2022/10/13/Different-flavors-Clang-compiler-Windows.html

I think the ``MD`` flag should be supported by both clang/clang-cl msvc-backed compilers, as they use the msvc runtime.

**Comment by jijjijj** [comment]
@memsharded
Thank you for the response.

I use CLang 19.1.3 provided by LLVM. I downloaded it right from their GitHub page. According to the article it should fall under the "MSVC-backed" clangs

![image](https://github.com/user-attachments/assets/7bbaacca-90ce-4e9e-bd84-e7e80099fabf)

**Comment by memsharded** [comment]
I am having a look.
These are the lines for ``clang-cl`` produced by CMake (for release and Debug respectively):
```
[1/2] C:\ws\LLVM\clang18\bin\clang-cl.exe  /nologo -TP  
-IC:\Users\conanws\kk\include /DWIN32 /D_WINDOWS /GR /EHsc /O2 /Ob2 /DNDEBUG
-MD -std:c++14
/showIncludes /FoCMakeFiles\mypkg.dir\src\mypkg.cpp.obj /FdCMakeFiles\mypkg.dir\mypkg.pdb 
-c -- C:\Users\conanws\kk\src\mypkg.cpp

[1/2] C:\ws\LLVM\clang18\bin\clang-cl.exe  /nologo -TP 
-IC:\Users\conanws\kk\include /DWIN32 /D_WINDOWS /GR /EHsc /Zi /Ob0 /Od /RTC1 
-MDd -std:c++14 
/showIncludes /FoCMakeFiles\mypkg.dir\src\mypkg.cpp.obj /FdCMakeFiles\mypkg.dir\mypkg.pdb
-c -- C:\Users\conanws\kk\src\mypkg.cpp
```

And with ``clang``:

```
[1/2] C:\ws\LLVM\clang18\bin\clang++.exe  -IC:/Users/conanws/kk/include -O3 -DNDEBUG 
-D_DLL -D_MT -Xclang --dependent-lib=msvcrt -std=c++14 -MD -MT
CMakeFiles/mypkg.dir/src/mypkg.cpp.obj -MF CMakeFiles\mypkg.dir\src\mypkg.cpp.obj.d 
-o CMakeFiles/mypkg.dir/src/mypkg.cpp.obj -c C:/Users/conanws/kk/src/mypkg.cpp

[1/2] C:\ws\LLVM\clang18\bin\clang++.exe  -IC:/Users/conanws/kk/include -O0 -g -Xclang -gcodeview
-D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -std=c++14 -MD -MT
CMakeFiles/mypkg.dir/src/mypkg.cpp.obj -MF CMakeFiles\mypkg.dir\src\mypkg.cpp.obj.d 
-o CMakeFiles/mypkg.dir/src/mypkg.cpp.obj -c C:/Users/conanws/kk/src/mypkg.cpp
```

The ``-MD`` and ``-MT`` in clang have different meaning https://clang.llvm.org/docs/ClangCommandLineReference.html

**Comment by jijjijj** [comment]
Seems CMake uses -Xclang which does the following:
![image](https://github.com/user-attachments/assets/b07cc087-9fb3-4646-9196-f0cbcd6a645f)

And `-cc1` means:
https://stackoverflow.com/questions/8991662/what-does-the-clang-cc1-option-do

**Comment by jijjijj** [comment]
So, is this a bug in AutotoolsToolchain?

**Comment by memsharded** [comment]
Yes, yes, it looks there is a gap, maybe not only in AutotoolsToolchain, but lack of model for clang/clang-cl differences, as the current model is more focused on the binary compatibility than on the compiler frontend input args.

**Comment by bopt-woba** [comment]
Is there any progress or decision yet? My development activities are also affected by this bug.

**Comment by memsharded** [comment]
I have started to work on https://github.com/conan-io/conan/pull/18347/files to provide tests for this.

The test ``test_autotools_bash_complete_clang()`` works both for ``clang`` and ``clang-cl``. It is true that the later does a warning for the ``-std=c++17`` flag, we need to have a look at this. But it is not failing the build, and the created binary looks correct. It will be great to have something that fails because of the flags differences, or at least warns about those flags:

```
conanfile.py: RUN: make -j8
conanvcvars.bat: Activating environment Visual Studio 17 - amd64 - winsdk_version=None - vcvars_ver=14.4
[vcvarsall.bat] Environment initialized for: 'x64'
source='main.cpp' object='main.obj' libtool=no \
DEPDIR=.deps depmode=msvc7msys /bin/sh ./depcomp \
/c/ws/llvm/18.1/bin/clang-cl -DPACKAGE_NAME=\"main\" -DPACKAGE_TARNAME=\"main\" -DPACKAGE_VERSION=\"1.0\" -DPACKAGE_STRING=\"main\ 1.0\" -DPACKAGE_BUGREPORT=\"some@email.com\" -DPACKAGE_URL=\"\" -DPACKAGE=\"main\" -DVERSION=\"1.0\" -I.   -DNDEBUG  -std=c++14 -MD -O2 -c -o main.obj `cygpath -w 'main.cpp'`
/c/ws/llvm/18.1/bin/clang-cl  -std=c++14 -MD -O2   -o main.exe main.obj  
clang-cl: warning: unknown argument ignored in clang-cl: '-std=c++14' [-Wunknown-argument]
clang-cl: warning: unknown argument ignored in clang-cl: '-std=c++14' [-Wunknown-argument]
```
